0018-logging.adoc

```diff
@@ -0,0 +1,94 @@
+* Dynamically controlling log level
+* Controlling the log level of libraries vs cnd logs i.e., 
+
+The output of this spike should be proposed changes to the current logging libraries (i.e., what libraries we use to log) and also some guidelines so the whole team knows when and where to add log statements.
+
+== Research
+
+=== Log Levels ===
+
+
+This is actually the easiest thing to solve, using correct log levels is pretty standard dev practice, the internet more or less has consensus on the following log levels:
+
+Copying from the following site: https://reflectoring.io/logging-levels/
+
+* error: should only be used when the application really is in trouble. Users are being affected without having a way to work around the issue.
+* warn: should be used when something bad happened, but the application still has the chance to heal itself or the issue can wait a day or two to be fixed.
+* info: should be used to document state changes in the application or some entity within the application.
+* debug: should be used to log any information that helps us identify what went wrong on DEBUG level.
+* trace: log all information that helps us to trace the flow of the application.
+
+=== Useful Context ===
+
+Useful context is not that easy to solve due to the async nature of cnd.
+Typically one can follow the sequence of log messages to get more context than is available within the scope of the log statement.
+The solution to this is most likely to use a tracing library instead of logging, see below for more on tracing.
+
+Related to context is useful output in the log messages i.e., don't dump byte slices using "{:?}".
+
+Suggestions to remedy this include:
+* Implement custom `Debug` for any objects that we would like to print within logs.
+* Add a `Log` trait for any objects that we wish to log and call the method when logging i.e., `log::info("Got: {}", foo.log())`
+
+=== Controlling Log Levels ===
+
+Currently we can control the base log level within the config file.  More fine grained control requires code changes in `cnd/src/logging/initialize.rs`.
+If we stay with logging I would suggest that we set the cnd log level to `trace`.
+It may be useful to have a few simple scripts in tree that do typical useful grep commands on the logs (unnecessary if we go with tracing).
+
+=== Tracing ===
+
+As mentioned in the issue on GitHub (linked above), logging may not be the correct approach because of the asynchronous nature of cnd.
+The issue is that because logging is event driven and the logs are ordered so it is hard to follow, from reading the log, what events are linked to each other.
+Tracing may be the solution to this but adding tracing requires a non-trivial implementation effort as well as some time investment for all devs on the team to learn how to debug using the trace output and interact with the tracing system.
+On top of this, tracing is most useful for, as the name suggests, 'tracing' a running application.
```

Comment:
The most useful aspect about `tracing` is that we can add context to the messages and therefore "build up" the context as the application is running. This means, each layer only needs to add _new information_ to the message because it inherits everything else. Best example for this is the swap id, once set, all the future messages for this one execution path have this swap id. So there is no need to repeat it.

Reply:
Agree 100%, will add this to the spike.